### PR TITLE
fix(compiler): panic when passing unknown keywords or wrong number of args 

### LINKED ISSUE #2244: Compiler panic when passing unknown keyword args
### I tried this

```js
class GitHub {
  init() {}
}

let gh = new GitHub(token: "hello");
```

### I expected this:

To give me a nice and friendly error

### Instead, this happened

```
thread '' panicked at 'called `Option::unwrap()` on a `None` value', libs/wingc/src/type_check.rs:1298:36
stack backtrace:
RuntimeError: unreachable
    at std::panicking::rust_panic_with_hook::h911ba172265603df (wasm://wasm/0094e3b2:wasm-function[210]:0xc076e)
    at core::panicking::panic_fmt::h6958314cbbe045b2 (wasm://wasm/0094e3b2:wasm-function[1399]:0x169e36)
    at core::panicking::panic::ha6e963274fcf290b (wasm://wasm/0094e3b2:wasm-function[1308]:0x1669d8)
    at wingc::type_check::TypeChecker::type_check_exp::hedc67d23352f6fd7 (wasm://wasm/0094e3b2:wasm-function[23]:0x311ce)
    at wingc::type_check::TypeChecker::type_check_scope::heb4270fd7820aade (wasm://wasm/0094e3b2:wasm-function[390]:0xff4f9)
    at wingc::type_check::h09744ad52f9a8d67 (wasm://wasm/0094e3b2:wasm-function[107]:0x95a45)
    at wingc::compile::h7291177120f9513e (wasm://wasm/0094e3b2:wasm-function[63]:0x73d74)
    at wingc_compile.command_export (wasm://wasm/0094e3b2:wasm-function[1538]:0x1705ac)
    at Object.invoke (fil[e:///home/httpsawingplaygroundvercelapp-mana/node_modules/winglang/dist/index.js:55511:37](vscode://file/e:///home/httpsawingplaygroundvercelapp-mana/node_modules/winglang/dist/index.js:55511:37))
    at compile (fil[e:///home/httpsawingplaygroundvercelapp-mana/node_modules/winglang/dist/index.js:54751:38](vscode://file/e:///home/httpsawingplaygroundvercelapp-mana/node_modules/winglang/dist/index.js:54751:38))
```

### Is there a workaround?

_No response_

### Component

Compiler

### Wing Version

_No response_

### Wing Console Version

_No response_

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### PR DESCRIPTION
**Pull request by attias**
Fixes #2244

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: libs/wingc/src/type_check.rs [hunk 1]
```diff
@@ -1337,14 +1345,35 @@ impl<'a> TypeChecker<'a> {
...
+						);
+					return self.expr_error(exp, err_text);
+				}
+
 				if !arg_list.named_args.is_empty() {
-					let last_arg = constructor_sig.parameters.last().unwrap().maybe_unwrap_option();
+					let last_arg = match constructor_sig.parameters.last() {
+						Some(arg) => arg.maybe_unwrap_option(),
+						None => {
+							return self.expr_error(exp, format!("Expected 0 named arguments when instantiating \"{}\"",
```
Comment (MarkMcCulloh) [inline libs/wingc/src/type_check.rs:1363]:
```suggestion
							return self.expr_error(exp, format!("Expected named arguments when instantiating \"{}\"",
```

File: libs/wingc/src/type_check.rs [hunk 2]
```diff
@@ -1337,14 +1345,35 @@ impl<'a> TypeChecker<'a> {
 				// Verify return type (This should never fail since we define the constructors return type during AST building)
 				self.validate_type(constructor_sig.return_type, type_, exp);
 
+				// Verify arity
+				let pos_args_count = arg_list.pos_args.len();
+				let min_args = constructor_sig.min_parameters();
+				if pos_args_count < min_args {
+					let err_text = format!(
+							"Expected {} required positional arguments but got {} when instantiating \"{}\"",
+							min_args, pos_args_count, type_
```
Comment (MarkMcCulloh) [inline libs/wingc/src/type_check.rs:1354]:
```suggestion
							"Expected {} arguments, but got {} when instantiating \"{}\"",
							min_args, pos_args_count, type_
```

File: libs/wingc/src/type_check.rs [hunk 3]
```diff
@@ -1424,8 +1453,30 @@ impl<'a> TypeChecker<'a> {
 					);
 				}
 
+				// Verify arity
+				let pos_args_count = arg_list.pos_args.len();
+				let min_args = func_sig.min_parameters();
+				if pos_args_count < min_args {
+					let err_text = format!(
+							"Expected {} required positional arguments but got {}",
```
Comment (MarkMcCulloh) [inline libs/wingc/src/type_check.rs:1461]:
```suggestion
							"Expected {} arguments but got {}",
```

File: libs/wingc/src/type_check.rs [hunk 4]
```diff
@@ -1424,8 +1453,30 @@ impl<'a> TypeChecker<'a> {
...
+						);
+					return self.expr_error(exp, err_text);
+				}
+
 				if !arg_list.named_args.is_empty() {
-					let last_arg = func_sig.parameters.last().unwrap().maybe_unwrap_option();
+					let last_arg = match func_sig.parameters.last() {
+						Some(arg) => arg.maybe_unwrap_option(),
+						None => {
+							return self.expr_error(exp, format!("Expected 0 named arguments for func at {}", exp.span().to_string()
```
Comment (MarkMcCulloh) [inline libs/wingc/src/type_check.rs:1471]:
```suggestion
							return self.expr_error(exp, format!("Expected named arguments"
```

File: libs/wingc/src/type_check.rs [hunk 5]
```diff
@@ -1424,8 +1453,30 @@ impl<'a> TypeChecker<'a> {
...
+							));
+						},
+					};
+
+					if !last_arg.is_struct() {
+						return self.expr_error(exp, format!("No named arguments expected"));
```
Comment (MarkMcCulloh) [inline libs/wingc/src/type_check.rs:1477]:
```suggestion
						return self.expr_error(exp, format!("Unexpected named arguments"));
```

File: examples/tests/invalid/class.w [hunk 6]
```diff
@@ -7,6 +7,22 @@ class C2 {}
...
+// ^^^^^^^^^ Expected 0 named arguments when instantiating \\"C4\\"
+
+class C5 {
+  init(foo: str)
+}
+new C5(); 
+// ^^^^^^^^^ Expected 1 required positional arguments but got 0 when instantiating "C5"
+
+new C5(foo: "bar"); 
+// ^^^^^^^^^ Expected 1 required positional arguments but got 0 when instantiating "C5"
```
Comment (eladb) [inline examples/tests/invalid/class.w:21]:
```suggestion
// ^^^^^^^^^ Expected 1 positional arguments but got 0
```

File: examples/tests/invalid/class.w [hunk 7]
```diff
@@ -7,6 +7,22 @@ class C2 {}
 new C2(1);
 // ^^^^^^^^^ Expected 0 arguments but got 1 when instantiating "C2"
```
Comment (eladb) [inline examples/tests/invalid/class.w:8]:
```suggestion
// ^^^^^^^^^ Expected 0 argument(s) but got 1
```

File: examples/tests/invalid/class.w [hunk 8]
```diff
@@ -7,6 +7,22 @@ class C2 {}
...
 
+class C4 {}
+new C4(token: "1");
+// ^^^^^^^^^ Expected 0 named arguments when instantiating \\"C4\\"
```
Comment (eladb) [inline examples/tests/invalid/class.w:12]:
```suggestion
// ^^^^^^^^^ Expected 0 named argument(s)"
```

**Review by eladb — CHANGES_REQUESTED** [review]
@revitalbarletz can you please help with this?

**Comment by attias** [comment]
@eladb please review

File: examples/tests/invalid/class.w [hunk 9]
```diff
@@ -5,7 +5,23 @@ class C1 {
...
+class C5 {
+  init(foo: str)
+}
+new C5(); 
+// ^^^^^^^^^ Expected 1 required positional arguments but got 0 when instantiating "C5"
+
+new C5(foo: "bar"); 
+// ^^^^^^^^^ Expected 1 positional argument(s) but got 0
+
+new C5("hello", foo: "bar");
```
Comment (yoav-steinberg) [inline examples/tests/invalid/class.w:23]: Should we also add tests for the modified logic in functions calls (not only class instantiation)?

**Review by eladb — APPROVED** [review]

**Review by eladb — APPROVED** [review]

**Review by eladb — APPROVED** [review]

**Pull request merged**